### PR TITLE
infer: fix Linux build and usage with macOS 11.3+ SDK

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

opam depexts is not aware of Linuxbrew, so it fails due to checking `apt` and other default package managers.

Part of this may be due to feature where opam depexts can access package manager to install missing dependencies.

Using `--no-depexts` or `--assume-depexts` is probably reasonable as missing dependencies will still cause failures later on and we don't need/want feature where opam runs `brew` commands.

https://github.com/Homebrew/homebrew-core/runs/3379835131?check_suite_focus=true
```
installing infer dependencies; this can take up to 30 minutes... 
Error:  Package conflict!
  * Missing dependency:
    - conf-autoconf
    depends on the unavailable system package 'autoconf'. Use `--no-depexts' to attempt installation anyway, or it is possible that a depext package name in the opam file is incorrect.
```

### CODE REVIEW

File: Formula/infer.rb [hunk 1]
```diff
@@ -70,6 +70,10 @@ def install
     # Release build
     touch ".release"
 
+    # Disable handling external dependencies as opam is not aware of Homebrew on Linux.
+    # Error:  Package conflict!  * Missing dependency:  - conf-autoconf
+    on_linux { inreplace "build-infer.sh", "infer \"$INFER_ROOT\" $locked", "\\0 --no-depexts" }
```
Comment (carlocab) [inline Formula/infer.rb:75]: We should upstream a patch that allows us to pass this flag.

**Comment by cho-m** [comment]
Will add some additional libraries based on macOS bottle. Need to check if they are linked on Linux afterward.
```
/usr/local/opt/infer/lib/infer/facebook-clang-plugins/clang/install/lib/libLLVM.dylib:
	/usr/local/opt/infer/lib/infer/facebook-clang-plugins/clang/install/lib/libLLVM.dylib (compatibility version 1.0.0, current version 11.1.0)
	/usr/lib/libedit.3.dylib (compatibility version 2.0.0, current version 3.0.0)
	/usr/lib/libz.1.dylib (compatibility version 1.0.0, current version 1.2.11)
	/usr/lib/libncurses.5.4.dylib (compatibility version 5.4.0, current version 5.4.0)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1292.60.1)
	/usr/lib/libxml2.2.dylib (compatibility version 10.0.0, current version 10.9.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 904.4.0)
```

**Comment by cho-m** [comment]
Don't know why Big Sur is failing. All changes are Linux or syntax-only, so shouldn't impact macOS build.

Linux failure seems to be that CMake couldn't find linker? Need to look further into this.
```
[763/764] Linking CXX static library /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a
FAILED: /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a 
: && /home/linuxbrew/.linuxbrew/Cellar/cmake/3.21.1/bin/cmake -E rm -f /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a && /tmp/clang-setup.dXPogM/build/bin/llvm-ar qc /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a  lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerCrossOver.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerDataFlowTrace.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerDriver.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerExtFunctionsDlsym.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerExtFunctionsWeak.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerExtFunctionsWindows.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerExtraCounters.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerFork.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerIO.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerIOPosix.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerIOWindows.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerLoop.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerMerge.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerMutate.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerSHA1.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerTracePC.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerUtil.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerUtilDarwin.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerUtilFuchsia.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerUtilLinux.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerUtilPosix.cpp.o lib/fuzzer/CMakeFiles/RTfuzzer.x86_64.dir/FuzzerUtilWindows.cpp.o && /tmp/clang-setup.dXPogM/build/bin/llvm-ranlib /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a && cd /tmp/clang-setup.dXPogM/build/tools/clang/runtime/compiler-rt-bins/lib/fuzzer/cxx_x86_64_merge.dir && CMAKE_LINKER-NOTFOUND --whole-archive /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a --no-whole-archive /tmp/clang-setup.dXPogM/build/tools/clang/runtime/compiler-rt-bins/lib/fuzzer/libcxx_fuzzer_x86_64/lib/libc++.a -r -o fuzzer_no_main.o && /tmp/clang-setup.dXPogM/build/bin/llvm-objcopy --localize-hidden fuzzer_no_main.o && /home/linuxbrew/.linuxbrew/Cellar/cmake/3.21.1/bin/cmake -E remove /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a && /tmp/clang-setup.dXPogM/build/bin/llvm-ar qcs /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer_no_main-x86_64.a fuzzer_no_main.o
/bin/sh: 1: CMAKE_LINKER-NOTFOUND: not found
[764/764] Linking CXX static library /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer-x86_64.a
FAILED: /tmp/clang-setup.dXPogM/build/lib/clang/11.1.0/lib/linux/libclang_rt.fuzzer-x86_64.a 
```

File: Formula/infer.rb (hunk 1 above)
Comment (cho-m) [reply Formula/infer.rb:77]: Will try to work on separate upstream enhancement.

Upstream `master` build script has modified this specific line, so we need different fixes. If I remember correctly, `inreplace` should fail when it is unable to find pattern, so this should fail on next version bump.

Maybe just something like following:
```diff
diff --git a/build-infer.sh b/build-infer.sh
index 757d40293..2dd9a3076 100755
--- a/build-infer.sh
+++ b/build-infer.sh
@@ -21,6 +21,7 @@ INFER_OPAM_DEFAULT_SWITCH="4.12.0+flambda"
 INFER_OPAM_DEFAULT_SWITCH_OPTIONS="--package=ocaml-variants.4.12.0+options,ocaml-option-flambda"
 INFER_OPAM_SWITCH=${INFER_OPAM_SWITCH:-$INFER_OPAM_DEFAULT_SWITCH}
 INFER_OPAM_SWITCH_OPTIONS=${INFER_OPAM_SWITCH_OPTIONS:-$INFER_OPAM_DEFAULT_SWITCH_OPTIONS}
+INFER_OPAM_INSTALL_DEPS_OPTIONS=${INFER_OPAM_INSTALL_DEPS_OPTIONS:-""}
 PLUGIN_DIR="$INFER_ROOT/facebook-clang-plugins"
 PLUGIN_SETUP_SCRIPT=${PLUGIN_SETUP_SCRIPT:-setup.sh}
 PLUGIN_SETUP="${PLUGIN_DIR}/clang/${PLUGIN_SETUP_SCRIPT}"
@@ -145,7 +146,7 @@ install_opam_deps () {
     if [ "$USE_OPAM_LOCK" == yes ]; then
         locked=.locked
     fi
-    opam install --deps-only "$INFER_ROOT"/opam/infer.opam$locked &&
+    opam install --deps-only "$INFER_ROOT"/opam/infer.opam$locked "$INFER_OPAM_INSTALL_DEPS_OPTIONS" &&
     if [ -n "$SANDCASTLE" ]; then
         opam pin list | grep yojson || opam pin add yojson "${DEPENDENCIES_DIR}/yojson-1.7.0fix"
     fi
```
Along with Formula change to `ENV["INFER_OPAM_INSTALL_DEPS_OPTIONS"] = "--no-depexts"`

**Comment by danielnachun** [comment]
The Linux failure now is probably because it needs `elfutils`:
```
/tmp/infer-20210821-12290-1gfjey5/infer-1.1.0/facebook-clang-plugins/clang/src/download/llvm-project/openmp/libomptarget/plugins/generic-elf-64bit/src/rtl.cpp:19:18: fatal error: gelf.h: No such file or directory
```

**Comment by cho-m** [comment]
> The Linux failure now is probably because it needs `elfutils`:

Due to `llvm` dependency, I think we need to use `libelf` rather than `elfutils`. Though the `libelf` formula is possibly dead/deprecated given we use archive.org and mirror URLs.

On side note, the `conflicts_with` makes it difficult to use `elfutils` right now as `libelf` is the more popular dependency (`llvm*`, `dwarfutils`, `mesa`, etc).

File: Formula/infer.rb [hunk 2]
```diff
@@ -32,11 +32,17 @@ class Infer < Formula
...
+  uses_from_macos "libxml2"
   uses_from_macos "ncurses"
   uses_from_macos "xz"
   uses_from_macos "zlib"
 
   on_linux do
+    # Add `llvm` for lld due to CMake bug where CC=clang doesn't fallback to ld.
+    # This causes error: /bin/sh: 1: CMAKE_LINKER-NOTFOUND: not found
+    # CMake PR ref: https://gitlab.kitware.com/cmake/cmake/-/merge_requests/6457
+    depends_on "llvm" => :build # TODO: remove when `cmake` is fixed
```
Comment (carlocab) [inline Formula/infer.rb:45]: This can be
```ruby
uses_from_macos "llvm" => :build
```

**Comment by cho-m** [comment]
Also on `elfutils` vs. `libelf`, do we want to consider migrating dependents to `elfutils`, and then deprecate `libelf`? And if there are still dependents, then maybe set as `keg_only` instead of `conflicts_with`?

For example, Arch Linux package https://archlinux.org/packages/core/x86_64/libelf/ uses `elfutils` version of `libelf`, which is why it is v0.185.

**Comment by danielnachun** [comment]
```
[*ERROR**][76319] *** stderr:
[*ERROR**][76319] FileUtils.cpp: In function ‘std::__cxx11::string FileUtils::makeAbsolutePath(const string&, std::__cxx11::string)’:
[*ERROR**][76319] FileUtils.cpp:44:57: error: invalid initialization of reference of type ‘const string& {aka const std::__cxx11::basic_string<char>&}’ from expression of type ‘llvm::StringRef’
[*ERROR**][76319]      const std::string &element(llvm::sys::path::filename(path));
```

I think this means we need to use a newer GCC than GCC 5.

**Comment by cho-m** [comment]
In this case, if I do `fails_with gcc: "5"`, would it use `llvm`? i.e. building `llvm` with `llvm`.

Though, if `gcc` is pulled in as dependency, it may end up using that.

**Comment by danielnachun** [comment]
> In this case, if I do `fails_with gcc: "5"`, would it use `llvm`? i.e. building `llvm` with `llvm`.
> 
> Though, if `gcc` is pulled in as dependency, it may end up using that.

No, it will only use `llvm` if you uses `fails_with :gcc`, which we shouldn't try yet.  The confusing thing here to remember is that `depends_on` only guarantees that a compiler is installed, but has no effect on what actually gets used.  It is the `fails_with` commands that force `brew` to actually use a different compiler.

**Comment by cho-m** [comment]
Okay. I'll try to add `gcc` for build. I may need to check on linkage since the LLVM `infer` builds should hopefully not have a dependency on `gcc` or `llvm` formulae.

Homebrew's `llvm` formula does build with `gcc-5` with some patches. I don't know what release `infer`'s LLVM is forked from.

**Comment by danielnachun** [comment]
It's forked from LLVM 11, which built fine with `gcc-5`, but I have no idea what they changed in their fork (one of many reasons why all these project should just upstream their changes to LLVM instead of forking), so it's possible that this fork of LLVM does need a newer GCC.

**Comment by cho-m** [comment]
Same failure with GCC-11. I don't know why.

I think upstream uses GCC for their own builds and dependency list includes it https://github.com/facebook/infer/blob/master/INSTALL.md#infer-dependencies-for-linux
> gcc >= 5.X or clang >= 3.4 (only needed for the C/Objective-C analysis)

I can try clang/llvm, but I can't tell if that would change anything.

**Comment by danielnachun** [comment]
This is actually an upstream bug that should be fixed by https://github.com/facebook/infer/commit/c90ec0683456e0f03135e7c059a1233351440736.  Let's apply this commit and go back to `gcc-5`.

**Comment by cho-m** [comment]
Finally built on Linux. Also, Big Sur keeps failing, but I don't think any of changes should be reason.

Remaining Linux errors include some missing dependencies for linkage like `libffi`:
```
==> brew linkage --test infer
==> FAILED
Missing libraries:
  unexpected (libffi.so.7)
```

And modifying test to use path to `javac`. The current test works on macOS due to the default wrappers provided:
```
==> /home/linuxbrew/.linuxbrew/Cellar/infer/1.1.0/bin/infer --fail-on-issue -P -- javac FailingTest.java
Capturing in javac mode...
Usage Error: 
*** Failed to execute compilation command: exited with code 127
*** Command: javac '@/tmp/infer-test-20210822-92093-ztxjg6/infer-out/tmp/javac_args.tmp.c94c0e'
*** Output:
/bin/sh: 1: javac: not found

*** Infer needs a working compilation command to run.
```

**Comment by cho-m** [comment]
Big Sur may be related to https://github.com/facebook/infer/issues/1439, relating to stdlib incompatibilities in macOS 11.3+.

This also impacts runtime behavior of existing bottles.

So, probably can ignore current Big Sur failures shouldn't be due to Linux-specific changes and is outside the scope of this PR.

**Comment by cho-m** [comment]
Linux changes are ready to review with CI passing.

I'll look into adding a way of passing `--no-depexts` as argument upstream. 
For now, the `inreplace` should work for current PR and is probably going to fail on next version bump (i.e. there is an upstream change on specific line being inreplace'd, so it is guaranteed to fail on `inreplace` or `patch` attempt). 

At that time, we can either switch to an environment variable / argument if available, or update the `inreplace`.

**Comment by danielnachun** [comment]
I think we could fix this on Big Sur if we can get this patch to apply: https://raw.githubusercontent.com/Homebrew/formula-patches/1f6960faf59a8d3d83ba8c32d0ec389febfee792/llvm%409/llvm%409.patch.  The SDK in 11.3 started including some files with conflicting names with the C++ standard libs that ship with LLVM, and that leads to these types of errors.  Would you mind trying?  I know this PR has already been a pain but if it works we'd save ourselves the trouble of fixing this later.

**Comment by cho-m** [comment]
That may be tricky as the Clang source is not in the `infer` repo. I think they download a copy of LLVM source and then apply patches on top of it. I have to look further into build script, but it may be https://github.com/facebook/infer/blob/v1.1.0/facebook-clang-plugins/clang/setup.sh#L15

We may need to see if we can embed our own patch file into the build.

**Comment by danielnachun** [comment]
> * add patch as resource block, copy to directory, and then patch/inreplace the setup script.

Let's try this option.  We should also open something upstream about this.  This type of problem is why they need to submit their patches to LLVM and maintain compatibility with the latest version of LLVM.

**Comment by carlocab** [comment]
I think you should be able to use a `patch` block directly by specifying the `strip` and `directory` options.

**Comment by cho-m** [comment]
Found an upstream commit to fix this: https://github.com/facebook/infer/commit/ec976d3be4e78dbbb019b3be941066f74e826880

**Review by danielnachun — APPROVED** [review]
I think we should merge as is and address the remaining issues in a follow up.

**Pull request closed without merging**